crates/ra_hir_def/src/item_scope.rs

```diff
@@ -7,13 +7,15 @@ use ra_db::CrateId;
 use rustc_hash::FxHashMap;
 
 use crate::{
-    db::DefDatabase, per_ns::PerNs, visibility::Visibility, AdtId, BuiltinType, HasModule, ImplId,
-    Lookup, MacroDefId, ModuleDefId, TraitId,
+    db::DefDatabase,
+    per_ns::{Compressed, PerNs},
+    visibility::Visibility,
+    AdtId, BuiltinType, HasModule, ImplId, Lookup, MacroDefId, ModuleDefId, TraitId,
 };
 
 #[derive(Debug, Default, PartialEq, Eq)]
 pub struct ItemScope {
-    visible: FxHashMap<Name, PerNs>,
+    visible: FxHashMap<Name, Compressed>,
```

Comment:
An alternative that comes to mind is storing three hash maps instead of just one. Also, there's a long-standing desire to library-ify name reoslution... So maybe it makes sense to think about the optimal design first? I feel that gradient descent-like local optimizations might not bring us to a sufficiently good optimium :)